### ISSUE TITLE
Bug with delete_versions/_recreate_virtual_dataset with h5py version 3.7

### ISSUE THREAD

**Issue by ArvidJB** [deshaw/versioned-hdf5#234]
I am running into a bug with the latest versioned_hdf5 code combine with h5py version 3.7 (this works with 3.6).
```
from versioned_hdf5 import VersionedHDF5File
import h5py
import numpy as np
from versioned_hdf5.replay import delete_versions

if __name__ == '__main__':
    with h5py.File('foo.h5', 'w') as f:
        vf = VersionedHDF5File(f)
        with vf.stage_version('r0') as sv:
            g = sv.create_group('data')
            dt = h5py.string_dtype(encoding='ascii')
            g.create_dataset('foo', data=['foo', 'bar'], dtype=dt)

    for i in range(1, 7):
        with h5py.File('foo.h5', 'r+') as f:
            vf = VersionedHDF5File(f)
            with vf.stage_version(f'r{i}') as sv:
                sv['data/foo'] = np.array([f'foo{i}', f'bar{i}'], dtype='O')

    with h5py.File('foo.h5', 'r+') as f:
        delete_versions(f, ['r1'])
```
which fails with the following error
```
Traceback (most recent call last):
  File "/codemill/bessen/fork/base/src/python/delete_versions_empty.py", line 22, in <module>
    delete_versions(f, ['r1'])
  File "/u/bessen/opensource/versioned-hdf5/versioned_hdf5/replay.py", line 378, in delete_versions
    delete_dataset(name)
  File "/u/bessen/opensource/versioned-hdf5/versioned_hdf5/replay.py", line 362, in delete_dataset
    _recreate_virtual_dataset(f, name, versions_to_keep, raw_data_chunks_map)
  File "/u/bessen/opensource/versioned-hdf5/versioned_hdf5/replay.py", line 300, in _recreate_virtual_dataset
    tmp_dataset = group.create_virtual_dataset(tmp_path, layout, fillvalue=dataset.fillvalue)
  File "/usr/local/python/python-3.10/std/lib64/python3.10/site-packages/h5py/_hl/group.py", line 193, in create_virtual_dataset
    dset = dataset.Dataset(dsid)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "/usr/local/python/python-3.10/std/lib64/python3.10/site-packages/h5py/_hl/dataset.py", line 622, in __init__
    self._dcpl = self.id.get_create_plist()
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5d.pyx", line 334, in h5py.h5d.DatasetID.get_create_plist
RuntimeError: Can't get creation plist (address of object past end of allocation)
```

I think this only fails for `dtype='object'`, but I am not sure. It looks like the `src_space` in the virtual_sources of the Dataset points to an entry past the allocated raw_data.

**Comment by ArvidJB** [comment]
> It looks like the src_space in the virtual_sources of the Dataset points to an entry past the allocated raw_data.

Ignore that, I think that's not correct.

This could be related to https://github.com/h5py/h5py/issues/941#issuecomment-666616083 which was supposedly fixed by https://github.com/h5py/h5py/pull/2044 ? The error message seems to be the same.

**Comment by ArvidJB** [comment]
Actually, I think what's missing is this workaround to replace `fillvalue` for variable length strings in `write_dataset`:
https://github.com/deshaw/versioned-hdf5/blob/master/versioned_hdf5/backend.py#L98:~:text=if%20fillvalue%20is,%7D)%22)

The same code is not present in `_recreate_virtual_dataset`. I added it locally and this seems to fix the problem.

**Comment by asmeurer** [comment]
I'm only able to reproduce this on the branch from https://github.com/h5py/h5py/pull/2111 because otherwise I get a segfault. Is that what you're using?

I'm unclear exactly what you added to `_recreate_virtual_dataset`. The code you linked to doesn't actually do anything. It's just some error checking. 

It seems like this whole thing is a bug in h5py. After h5py commit 39d958a73d070a9f9606e935109c5147243e37df, the code starts to segfault. The PR https://github.com/h5py/h5py/pull/2111 fixes that segfault, but produces this error. So one of the changes in between changed the behavior here. 

I can fix it with 

```diff
diff --git a/versioned_hdf5/replay.py b/versioned_hdf5/replay.py
index 2fe3f07..e19ae4c 100644
--- a/versioned_hdf5/replay.py
+++ b/versioned_hdf5/replay.py
@@ -297,7 +297,8 @@ def _recreate_virtual_dataset(f, name, versions, raw_data_chunks_map, tmp=False)
         head, tail = pp.split(name)
         tmp_name = '_tmp_' + tail
         tmp_path = pp.join(head, tmp_name)
-        tmp_dataset = group.create_virtual_dataset(tmp_path, layout, fillvalue=dataset.fillvalue)
+        fillvalue = None if dataset.dtype == object else dataset.fillvalue
+        tmp_dataset = group.create_virtual_dataset(tmp_path, layout, fillvalue=fillvalue)

         for key, val in dataset.attrs.items():
             tmp_dataset.attrs[key] = val
```

It's not clear to me if this is fully correct. It's saying the fillvalue should be `None` for object dtypes (which is what it used to be for this dataset), but is it possible to have other types of object dtype datasets with other fillvalues?

**Comment by ArvidJB** [comment]
Yes, we are using the patch from https://github.com/h5py/h5py/pull/2111/files
I am not totally clear on what has changed in h5py version 3.7 which now causes this to be a problem.

I fixed our code by copying some code to override `fillvalue` for dtype='object' Datasets from `create_virtual_dataset` to `recreate_virtual_dataset` (that's essentially the same as your fix above):
```
diff --git a/versioned_hdf5/replay.py b/versioned_hdf5/replay.py
index 2fe3f07..215e931 100644
--- a/versioned_hdf5/replay.py
+++ b/versioned_hdf5/replay.py
@@ -297,7 +297,17 @@ def _recreate_virtual_dataset(f, name, versions, raw_data_chunks_map, tmp=False)
         head, tail = pp.split(name)
         tmp_name = '_tmp_' + tail
         tmp_path = pp.join(head, tmp_name)
-        tmp_dataset = group.create_virtual_dataset(tmp_path, layout, fillvalue=dataset.fillvalue)
+        dtype = raw_data.dtype
+        fillvalue = dataset.fillvalue
+        if dtype.metadata and ('vlen' in dtype.metadata or 'h5py_encoding' in dtype.metadata):
+            # Variable length string dtype
+            # (https://h5py.readthedocs.io/en/2.10.0/strings.html). Setting the
+            # fillvalue in this case doesn't work
+            # (https://github.com/h5py/h5py/issues/941).
+            if fillvalue not in [0, '', b'', None]:
+                raise ValueError("Non-default fillvalue not supported for variable length strings")
+            fillvalue = None
+        tmp_dataset = group.create_virtual_dataset(tmp_path, layout, fillvalue=fillvalue)

         for key, val in dataset.attrs.items():
             tmp_dataset.attrs[key] = val
```

I also agree that this is not 100% correct and will break if we ever have an actual `fillvalue` for object dtypes. But luckily we are not using that at the moment!

Long story short: the patch in [your comment](https://github.com/deshaw/versioned-hdf5/issues/234#issuecomment-1216046297) should resolve our problem.
